From c9bf835731863f8a1993f9b1806da7bc6100ce59 Mon Sep 17 00:00:00 2001
From: Bai Ping <ping.bai@nxp.com>
Date: Mon, 8 Jan 2018 16:06:55 +0800
Subject: [PATCH 3247/5242] MLK-17447 drivers: soc: imx: Fix busfreq mutex
 unlock twice on imx8mq

commit  ae5cd5b48cac99e305c50878899530a34ac3b04f from
https://source.codeaurora.org/external/imx/linux-imx.git

A 'return' statement is missed before, So the mutex will be unlocked
twice, in some corner case, one core will unlock the mutex that locked
by anohter core wrongly. Then lead to concurrent access to the DVFS
at the same time.

Signed-off-by: Bai Ping <ping.bai@nxp.com>
Reviewed-by: Anson Huang <Anson.Huang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/soc/imx/busfreq-imx8mq.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/soc/imx/busfreq-imx8mq.c b/drivers/soc/imx/busfreq-imx8mq.c
index cba7970..4d2ce9d 100644
--- a/drivers/soc/imx/busfreq-imx8mq.c
+++ b/drivers/soc/imx/busfreq-imx8mq.c
@@ -1,5 +1,5 @@
 /*
- * Copyright 2017 NXP
+ * Copyright 2017-2018 NXP
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License
@@ -330,6 +330,7 @@ void release_bus_freq(enum bus_freq_mode mode)
 		(audio_bus_count == 0)) {
 		set_low_bus_freq();
 		mutex_unlock(&bus_freq_mutex);
+		return;
 	}
 
 	mutex_unlock(&bus_freq_mutex);
-- 
1.7.9.5

